### PR TITLE
Fix: 4477 filtering by count on an aggregate

### LINKED ISSUE #4477: Nested aggregate filtering not working after updating neo4j-graphql from V3 to V4
I believe this is bug introduced between v3 and v4.

Given the following type defs:

```
  type Brand {
      services: [Service!]! @relationship(type: "HAS_SERVICE", direction: OUT)
  }

  type Collection {
      services: [Service!]! @relationship(type: "HAS_SERVICE", direction: OUT)
  }

  type Service {
      collection: Collection @relationship(type: "HAS_SERVICE", direction: IN)
  }
```

If I run a query as follows (to find Services that aren't in a Collection):
```
query ExampleQuery {
  brands {
    services(where: { collectionAggregate: { count: 0 } }) {
      collectionAggregate {
        count
      }
    }
  }
}
```
In V3 I got the following result (as expected):
```
{
  data: {
    "brands": [
      {
        "services": [
          {
            "collectionAggregate": {
              "count": 0
            }
          }
        ]
      }
    ]
  }
}
```
But now, in V4, I get this result: 
```
{
  data: {
    "brands": [
      {
        "services": [
          {
            "collectionAggregate": {
              "count": 0
            }
          },
          {
            "collectionAggregate": {
              "count": 1
            }
          }
        ]
      }
    ]
  }
}
```
So, in V4, the aggregate filter is not being respected. However, if I query with Services at the top level (as opposed to Brands), the aggregate filter works correctly, leading me to believe that the issue is only effecting aggregate filters when nested.

i.e. the following query, in V4, with the Services at top level, returns only Services that are not related to a Collection (as expected):
```
query ExampleQuery {
    services(where: { collectionAggregate: { count: 0 } }) {
      collectionAggregate {
        count
      }
    }
}
```

Thanks.

### PR DESCRIPTION
**Pull request by mjfwebb**
# Description

Filtering wasn't working properly on nested read operations, so this PR adds that in.

## Complexity

Low

There are several changes in `ReadOperation` but it is mainly code re-organisation to aid future refactoring

# Issue

Closes #4477

### CODE REVIEW

**Comment by neo4j-team-graphql** [comment]
Made some changes to the Neo4j GraphQL Toolbox? Or made some changes to the core library perhaps?

Check out your changes at https://neo4j-graphql-toolbox-4507.surge.sh

Use for instance a local dbms hosted on [Neo4j Desktop](https://neo4j.com/download/) or a cloud-hosted [Aura Free](https://neo4j.com/cloud/aura-free/) dbms.

**Comment by neo4j-team-graphql** [comment]
# Performance Report

No Performance Changes


<details>
    <summary>Show Full Table</summary>
    
| name | dbHits | old dbHits | time (ms) | old time (ms) | maxRows |
| ---- | ------ | ---------- | --------- | ------------- | ------- |
|  aggregations.TopLevelAggregate | 3404 | 3404 | 31 | 40 | 1134 |
|  aggregations.TopLevelAggregateWithMultipleFields | 6802 | 6802 | 40 | 51 | 1134 |
|  aggregations.NestedAggregation | 16544 | 16544 | 58 | 80 | 2174 |
|  aggregations.AggregationWithWhere | 11970 | 11970 | 37 | 51 | 2174 |
|  aggregations.AggregationWhereWithinNestedRelationships | 22116478 | 22116478 | 2188 | 2294 | 2008534 |
|  aggregations.AggregationWhereWithinNestedConnections | 22116478 | 22116478 | 2193 | 2259 | 2008534 |
|  aggregations.NestedCountFromMovieToActors | 9701 | 9701 | 38 | 36 | 2174 |
|  aggregations.NestedCountFromActorsToMovie | 9928 | 9928 | 33 | 40 | 2174 |
|  aggregations.DeeplyNestedCount | 11062184 | 11062184 | 2509 | 2923 | 2008534 |
|  aggregations.InterfacesAggregations | 6242 | 6242 | 66 | 52 | 2080 |
|  aggregations.InterfacesAggregationsWithTwoFields | 11444 | 11444 | 91 | 86 | 2080 |
|  batch-create.BatchCreate | 4200 | 4200 | 137 | 156 | 600 |
|  batch-create.BatchCreateSmall | 77 | 77 | 49 | 66 | 11 |
|  connect.createAndConnect | 14424 | 14424 | 139 | 192 | 3003 |
|  connections.Connection | 14049 | 14049 | 55 | 83 | 2174 |
|  connections.NestedConnection | 41068 | 41068 | 113 | 126 | 4516 |
|  create.SimpleMutation | 7 | 7 | 45 | 59 | 1 |
|  cypher-directive.TopLevelMutationDirective | 1135 | 1135 | 24 | 23 | 1134 |
|  delete.SimpleDelete | 19401 | 19401 | 187 | 193 | 1040 |
|  delete.NestedDeleteInUpdate | 19165 | 19165 | 129 | 175 | 1183 |
|  2871.NestedRelationshipFilter | 23082 | 23077 | 103 | 56 | 4434 |
|  2925.SingleRelationshipFilter | 6468 | 6468 | 39 | 40 | 1040 |
|  2925.NestedSingleRelationshipFilter | 22891 | 22891 | 72 | 86 | 2174 |
|  2925.SingleRelationshipRequiredFilter | 5201 | 5201 | 42 | 40 | 1040 |
|  2925.NestedSingleRelationshipRequiredFilter | 9361 | 9361 | 63 | 52 | 1040 |
|  query.SimpleQuery | 3121 | 3121 | 18 | 19 | 1040 |
|  query.SimpleQueryWithRelationship | 16129 | 16129 | 35 | 45 | 2174 |
|  query.QueryWhere | 9691 | 9676 | 30 | 41 | 2166 |
|  query.SimpleQueryWithNestedWhere | 9858 | 9843 | 46 | 56 | 2166 |
|  query.Nested | 10095502 | 10095502 | 6792 | 6715 | 2008534 |
|  query.NestedWithFilter | 10074351 | 10074351 | 6616 | 6428 | 2004000 |
|  query.OrFilterOnRelationships | 43661 | 43878 | 139 | 177 | 2033 |
|  query.OrFilterOnRelationshipsAndNested | 36744 | 36756 | 185 | 221 | 2033 |
|  query.QueryWithNestedIn | 13857 | 14010 | 47 | 62 | 1379 |
|  query.NestedConnectionWhere | 9801 | 9801 | 50 | 70 | 2174 |
|  query.DeeplyNestedConnectionWhere | 9893 | 9799 | 83 | 85 | 2174 |
|  query.DeeplyNestedWithRelationshipFilters | 18746 | 18686 | 113 | 141 | 1586 |
|  query.NestedWithRelationshipSingleFilters | 3865 | 3865 | 140 | 164 | 1134 |
|  query.Fulltext | 64 | 64 | 26 | 33 | 16 |
|  query.FulltextWithNestedQuery | 558 | 558 | 36 | 43 | 84 |
|  sorting-and-cypher.TopLevelSortWithCypher | 14059 | 14059 | 49 | 45 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithCypher | 14059 | 14059 | 59 | 65 | 2174 |
|  sorting-and-cypher.TopLevelSortWithCypherWithNested | 14185 | 14133 | 53 | 61 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithCypherWithNested | 14185 | 14133 | 107 | 111 | 2174 |
|  sorting-and-cypher.TopLevelSortWithExpensiveCypher | 14661 | 14536 | 90 | 120 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithExpensiveCypher | 14480 | 14461 | 98 | 103 | 2174 |
|  sorting.SortMultipleTypes | 3498 | 3498 | 72 | 82 | 1040 |
|  sorting.SortMultipleTypesWithCypherWithCypher | 14460 | 14453 | 94 | 111 | 2174 |
|  sorting.SortOnNestedFields | 14049 | 14049 | 48 | 56 | 2174 |
|  sorting.SortDeeplyNestedFields | 43148 | 43148 | 92 | 94 | 4516 |
|  sorting.ConnectionWithSort | 3281 | 3281 | 77 | 69 | 1040 |
|  unions.SimpleUnionQuery | 321 | 321 | 49 | 52 | 35 |
|  unions.SimpleUnionQueryWithMissingFields | 293 | 293 | 60 | 52 | 35 |
|  unions.NestedUnion | 410637 | 410637 | 317 | 248 | 33033 |
|  unions.NestedUnionWithMissingFields | 384611 | 384611 | 261 | 247 | 33033 |
|  update.NestedUpdate | 16143 | 16143 | 87 | 119 | 2002 |

</details>
        
Old Schema Generation: 24.032s
Schema Generation: 23.756s
Old Subgraph Schema Generation: 27.807s
Subgraph Schema Generation: 26.646s

File: .changeset/unlucky-forks-pretend.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"@neo4j/graphql": patch
+---
+
+Adds filtering on nested read operations
```
Comment (angrykoala) [inline .changeset/unlucky-forks-pretend.md:5]: Is this the correct changelog for this? 

It sounds like we are adding a feature, but this is a bugfix isn't it?

File: packages/graphql/src/translate/queryAST/ast/operations/ReadOperation.ts [hunk 2]
```diff
@@ -100,47 +100,127 @@ export class ReadOperation extends Operation {
...
+        const filterSubqueries = wrapSubqueriesInCypherCalls(context, this.filters, [context.target]);
+        const filterPredicates = this.getPredicates(context);
+        const fieldSubqueries = this.getFieldsSubqueries(context);
+        const cypherFieldSubqueries = this.getCypherFieldsSubqueries(context);
+        const subqueries = Cypher.concat(...fieldSubqueries);
+        const sortSubqueries = wrapSubqueriesInCypherCalls(context, this.sortFields, [context.target]);
+
+        // const authFilterSubqueries = this.getAuthFilterSubqueries(nestedContext).map((sq) =>
+        //     new Cypher.Call(sq).innerWith(nestedContext.target)
+        // );
```
Comment (angrykoala) [inline packages/graphql/src/translate/queryAST/ast/operations/ReadOperation.ts:119]: Commented code :eye:
```suggestion
```

File: packages/graphql/src/translate/queryAST/ast/operations/ReadOperation.ts [hunk 3]
```diff
@@ -100,47 +100,127 @@ export class ReadOperation extends Operation {
         return filterTruthy(this.authFilters.map((f) => f.getPredicate(context)));
     }
 
+    private transpileWithCreateSelection(context: QueryASTContext): OperationTranspileResult {
```
Comment (angrykoala) [inline packages/graphql/src/translate/queryAST/ast/operations/ReadOperation.ts:103]: If I understand correctly, here you have moved all the transpile for create operations flow to a separate method.

I'm not sure if that is the best approach for future refactors. As the idea would be to make the code the same between create and non-create, and having this as a separate flow leads to more 

Thoughts? Or maybe I'm missing something

Comment (mjfwebb) [reply packages/graphql/src/translate/queryAST/ast/operations/ReadOperation.ts]: This change has been removed from this PR now, and will be addressed in a future refactor

**Review by angrykoala — APPROVED** [review]

**Pull request merged**